drivers/gpio/gpio_eos_s3.c

```diff
@@ -0,0 +1,403 @@
+					   uint32_t mask,
+					   uint32_t value)
+{
+	ARG_UNUSED(dev);
+	uint32_t target_value;
+	uint32_t output_states = MISC_CTRL->IO_OUTPUT;
+
+	target_value = ((output_states & ~mask) | (value & mask));
+	MISC_CTRL->IO_OUTPUT = (target_value & GPIOS_MASK);
+
+	return 0;
+}
+
+static int gpio_eos_s3_port_set_bits_raw(const struct device *dev,
+					 uint32_t mask)
+{
+	ARG_UNUSED(dev);
+
+	MISC_CTRL->IO_OUTPUT |= (mask & GPIOS_MASK);
+
+	return 0;
+}
+
+static int gpio_eos_s3_port_clear_bits_raw(const struct device *dev,
+					   uint32_t mask)
+{
+	ARG_UNUSED(dev);
+
+	MISC_CTRL->IO_OUTPUT &= ~(mask & GPIOS_MASK);
+
+	return 0;
+}
+
+static int gpio_eos_s3_port_toggle_bits(const struct device *dev,
+					uint32_t mask)
+{
+	ARG_UNUSED(dev);
+	uint32_t target_value;
+	uint32_t output_states = MISC_CTRL->IO_OUTPUT;
+
+	target_value = output_states ^ mask;
+	MISC_CTRL->IO_OUTPUT = (target_value & GPIOS_MASK);
+
+	return 0;
+}
+
+static int gpio_eos_s3_manage_callback(const struct device *dev,
+				       struct gpio_callback *callback, bool set)
+{
+	struct gpio_eos_s3_data *data = dev->data;
+
+	return gpio_manage_callback(&data->callbacks, callback, set);
+}
+
+static int gpio_eos_s3_pin_interrupt_configure(const struct device *dev,
+					       gpio_pin_t gpio_num,
+					       enum gpio_int_mode mode,
+					       enum gpio_int_trig trig)
+{
+	struct gpio_eos_s3_data *data = dev->data;
+	GPIOCfgTypeDef gpio_cfg;
+
+	gpio_cfg.ucGpioNum = gpio_num;
+	gpio_cfg.xPadConf = gpio_eos_s3_pad_select(dev, gpio_num);
+
+	if (mode == GPIO_INT_MODE_DISABLED) {
+		/* Get IRQ number which should be disabled */
+		int irq_num = gpio_eos_s3_get_irq_num(gpio_cfg.xPadConf->ucPin);
+
+		if (irq_num < 0) {
+			return -EINVAL;
+		}
+
+		/* Disable IRQ */
+		INTR_CTRL->GPIO_INTR_EN_M4 &= ~BIT((uint32_t)irq_num);
+
+		/* Clear IRQ mask and corresponding IRQ number */
+		data->gpio_irq_mask &= ~BIT((uint32_t)irq_num);
+		data->gpio_irqs[irq_num] = DISABLED_GPIO_IRQ;
+
+		/* Clear configuration */
+		INTR_CTRL->GPIO_INTR_TYPE &= ~((uint32_t)(BIT(irq_num)));
+		INTR_CTRL->GPIO_INTR_POL &= ~((uint32_t)(BIT(irq_num)));
+	} else {
+		/* Prepare configuration */
+		if (mode == GPIO_INT_MODE_LEVEL) {
+			gpio_cfg.intr_type = LEVEL_TRIGGERED;
+			if (trig == GPIO_INT_TRIG_LOW) {
+				gpio_cfg.pol_type = FALL_LOW;
+			} else {
+				gpio_cfg.pol_type = RISE_HIGH;
+			}
+		} else {
+			gpio_cfg.intr_type = EDGE_TRIGGERED;
+			switch (trig) {
+			case GPIO_INT_TRIG_LOW:
+				gpio_cfg.pol_type = FALL_LOW;
+				break;
+			case GPIO_INT_TRIG_HIGH:
+				gpio_cfg.pol_type = RISE_HIGH;
+				break;
+			case GPIO_INT_TRIG_BOTH:
+				return -ENOTSUP;
+			}
+		}
+
+		/* Set IRQ configuration */
+		int irq_num = HAL_GPIO_IntrCfg(&gpio_cfg);
+
+		if (irq_num < 0) {
+			return -EINVAL;
+		}
+
+		data->gpio_irqs[irq_num] = gpio_num;
+
+		data->gpio_irq_mask |= BIT((uint32_t)irq_num);
+
+		/* Enable IRQ */
+		INTR_CTRL->GPIO_INTR_EN_M4 |= BIT((uint32_t)irq_num);
```

Comment:
I didn't find detailed information about interrupt controller in the datasheet, however we probably should clear the interrupt status register before enabling the interrupt. Something like
```
	INTR_CTRL->GPIO_INTR |= BIT((uint32_t)irq_num);
```
On many architectures if the status register was dirty and we enable the interrupt it will trigger immediately.

Reply:
I agree, this is something I missed and it is definitely worth adding - fixed.